[web-animations] keyframes should be recomputed if used CSS variable changes

graouts opened the pull request:
#### 5c0b3cb4b2450bcc60e21f512306aaaa56e3e9b3
<pre>
[web-animations] keyframes should be recomputed if used CSS variable changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=248145">https://bugs.webkit.org/show_bug.cgi?id=248145</a>

Reviewed by Antti Koivisto.

When the value for a CSS variable changes, we must ensure that any set of keyframes that use that CSS variable
are recomputed, whether the animation is a CSS Animation of a script-originated animation. This does not apply
to CSS Transitions which would operate on resolved values in RenderStyle.

To do this we add a StyleRuleKeyframe::containsCSSVariableReferences() method which indicates whether a keyframe
rule contains CSS variables. Then, we add a similar method on KeyframeEffect returning the boolean flag computed
resolving keyframes in KeyframeEffect::computeCSSAnimationBlendingKeyframes(), for the CSS Animations case, and
KeyframeEffect::updateBlendingKeyframes(), for the script-originated animation case.

Then in KeyframeEffectStack::applyKeyframeEffects(), much like we do for detecting changes made to font-size, we
check whether any CSS variable (or custom property in WebCore parlance) has changed and recompute keyframes if that
is the case.

We now pass the final two subtests in web-animations/animation-model/keyframe-effects/effect-value-context-filling.html
and since those tests only test the script-originated animation case, we also add a new test in css/css-animations
to test the CSS Animations case.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-css-variable-in-keyframe-adjusted-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-css-variable-in-keyframe-adjusted.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-context-filling-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::updateBlendingKeyframes):
(WebCore::KeyframeEffect::computeCSSAnimationBlendingKeyframes):
* Source/WebCore/animation/KeyframeEffect.h:
(WebCore::KeyframeEffect::containsCSSVariableReferences const):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyKeyframeEffects):
* Source/WebCore/css/CSSKeyframeRule.cpp:
(WebCore::StyleRuleKeyframe::containsCSSVariableReferences const):
* Source/WebCore/css/CSSKeyframeRule.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::keyframeStylesForAnimation):
* Source/WebCore/style/StyleResolver.h:

Canonical link: <a href="https://commits.webkit.org/256893@main">https://commits.webkit.org/256893@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56e65947539a1151fb99fca0c78d0169e7b8dd84

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97134 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6400 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106651 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166923 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101105 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6662 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35134 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89523 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103341 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102802 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5003 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83755 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31995 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86852 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88685 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74883 "Found 1 new API test failure: /WebKitGTK/TestWebKitUserContentFilterStore:/webkit/WebKitUserContentFilterStore/save-from-file (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/422 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/406 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21606 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5204 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44112 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2332 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1646 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40921 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -873,6 +873,9 @@ void KeyframeEffect::updateBlendingKeyframes(RenderStyle& elementStyle, const St
     auto& styleResolver = m_target->styleResolver();
 
+    if (!m_parsedKeyframes.isEmpty())
+        m_containsCSSVariableReferences = false;
```
Can't this just be unconditional?

graouts replied:
Actually, yes, I _think_ so.

anttijk commented on Source/WebCore/animation/KeyframeEffectStack.cpp:
```diff
@@ -171,8 +171,19 @@ OptionSet<AnimationImpact> KeyframeEffectStack::applyKeyframeEffects(RenderStyle
+                if (&previousLastStyleChangeEventStyle->inheritedCustomProperties() != &targetStyle.inheritedCustomProperties()
+                    || previousLastStyleChangeEventStyle->inheritedCustomProperties() != targetStyle.inheritedCustomProperties()
+                    || &previousLastStyleChangeEventStyle->nonInheritedCustomProperties() != &targetStyle.nonInheritedCustomProperties()
+                    || previousLastStyleChangeEventStyle->nonInheritedCustomProperties() != targetStyle.nonInheritedCustomProperties())
```
This is not quite right. If pointers are equivalent you don't need to test deep equivalence. Even with different pointers the deep comparison may still be equivalent. You want

```
if (&a != &b && a != b) 
   return true;
```

or both inherited and non inherited pairs.

graouts replied:
Wait, the very first check is for pointers _not_ being the same, which should be enough to determine there was a change, isn't it? If a custom property changed, could the pointers be equivalent still?

graouts replied:
Anyway, I've updated the patch with the check as suggested.

graouts committed:
[web-animations] keyframes should be recomputed if used CSS variable changes
https://bugs.webkit.org/show_bug.cgi?id=248145

Reviewed by Antti Koivisto.

When the value for a CSS variable changes, we must ensure that any set of keyframes that use that CSS variable
are recomputed, whether the animation is a CSS Animation of a script-originated animation. This does not apply
to CSS Transitions which would operate on resolved values in RenderStyle.

To do this we add a StyleRuleKeyframe::containsCSSVariableReferences() method which indicates whether a keyframe
rule contains CSS variables. Then, we add a similar method on KeyframeEffect returning the boolean flag computed
resolving keyframes in KeyframeEffect::computeCSSAnimationBlendingKeyframes(), for the CSS Animations case, and
KeyframeEffect::updateBlendingKeyframes(), for the script-originated animation case.

Then in KeyframeEffectStack::applyKeyframeEffects(), much like we do for detecting changes made to font-size, we
check whether any CSS variable (or custom property in WebCore parlance) has changed and recompute keyframes if that
is the case.

We now pass the final two subtests in web-animations/animation-model/keyframe-effects/effect-value-context-filling.html
and since those tests only test the script-originated animation case, we also add a new test in css/css-animations
to test the CSS Animations case.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-css-variable-in-keyframe-adjusted-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/animation-css-variable-in-keyframe-adjusted.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-context-filling-expected.txt:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::updateBlendingKeyframes):
(WebCore::KeyframeEffect::computeCSSAnimationBlendingKeyframes):
* Source/WebCore/animation/KeyframeEffect.h:
(WebCore::KeyframeEffect::containsCSSVariableReferences const):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyKeyframeEffects):
* Source/WebCore/css/CSSKeyframeRule.cpp:
(WebCore::StyleRuleKeyframe::containsCSSVariableReferences const):
* Source/WebCore/css/CSSKeyframeRule.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::keyframeStylesForAnimation):
* Source/WebCore/style/StyleResolver.h:

Canonical link: https://commits.webkit.org/256893@main
